enhancements/local-storage/health-monitoring.md

```diff
@@ -0,0 +1,142 @@
+
+## Proposal
+
+[node_exporter](https://github.com/openshift/node_exporter) currently exports per-node metrics, we will use it to collect and export metrics for each device.
+
+### User Stories
+
+- As an admin, I would like to be alerted when a storage device is unhealthy and be linked to remediation steps.
+- As an admin, I would like if components (such as OpenShift Data Foundation) could access storage device health and failover faster.
+
+### Implementation Details/Notes/Constraints
+
+Implement a sidecar in the [node_exporter](https://github.com/openshift/node_exporter) component under [Cluster Monitoring Operator](https://github.com/openshift/cluster-monitoring-operator) that will scrape the metrics from the node and export them via the [textfile exporter](https://github.com/openshift/node_exporter/tree/master/text_collectors) (an exporter that scans `.prom` files in the
+`/var/node_exporter/textfile` directory and exports them.
+
+The Cluster Monitoring Operator will deploy the sidecar when the platform is Baremetal or None by modifying the `node_exporter` [daemonset](https://github.com/openshift/cluster-monitoring-operator/blob/master/assets/node-exporter/daemonset.yaml). The platform will be detected by querying the `infrastructure` resource.
+
+After that, the maintainence effort will mostly centered on the code the image runs to get health and update its `.prom` file. The logic for this collector can be in a subdirectory of CMO and have it's own OWNERS file allowing the ODF team to own it.
+
+We will be collecting metrics via [libstoragemgmt](https://github.com/libstorage/libstoragemgmt) and calling it via [libstoragemgmt-golang](https://github.com/libstorage/libstoragemgmt-golang/).
+If a block device is a partion or logical volume, we will only expose health for the parent device.
+
+
+### Ownership and Maintenance
+
+- The `node_exporter` daemonset creation is owned by the Openshift-Monitoring team. The PR for modifying the daemonset based on infrastructure to include the sidecar can be contributed by ODF, but Monitoring team will have to review.
+- The sidecar image and logic will live in the Cluster Monitoring Operator repo, but maintenance and ownership will be with the ODF team via a OWNERS file.
+- A new bugzilla subcomponent will be created under openshift-monitoring, but the default assignee will be ODF and ODF will be responsible for triaging the issues.
+
+### Risks and Mitigations
+
+- The golang bindings for libstoragemgmgt need some review and polish.
+  - Mitigation: The ODF(OpenShift Data Foundation) team will contribute to and help the RHEL team maintain libstoragemgmt golang
+
+## Design Details
+
+Each device will have a metric `storage_device_health` metric associated with it that has a value of 0,1, or 2. Where 0 is healthy state, 1 is a warning state, and 2 is a failed state. Each metric will be labeled by node, device-name, and device-id (where available).
+
```

Comment:
When we've discussed adding other hardware-related metrics in the past there was some concern about the cardinality of the data and how much new data would need to be processed and stored by Prometheus. Do we have those sorts of concerns for this data?

Reply:
There are currently plans to add one new metric per storage device.

Reply:
It would be good to provide an estimation of how many timeseries would be exposed overall (e.g. `for a node with N physical storage devices, the exporter would expose (number of metrics) * (number of storage device) = 1*N = N timeseries`).

Reply:
Not sure if this is the right place, but should this proposal also mention how this health status is derived? Will the code look at SMART attributes or the result of SMART test runs or something else entirely?

If this would rely on SMART's testing feature for example, these tests need to be scheduled as well.

I think the implementation choice could inform whether using the textfile collector interface or a stand-alone exporter is preferable. Say, if the health status is determined based on SMART tests, a standalone exporter, that queries the device synchronously might be overly resource intensive, as running those tests is typically done fairly rarely.